File: Classwork/dataStructures/src/mazeAttempts/maze2/Maze.java
package mazeAttempts.maze2;

import org.jetbrains.annotations.NotNull;

import javax.swing.*;
import java.awt.*;
import java.awt.event.KeyEvent;
import java.awt.event.KeyListener;
import java.awt.event.MouseEvent;
import java.awt.event.MouseListener;
import java.io.BufferedReader;
import java.io.FileReader;
import java.util.ArrayList;

import static java.awt.Color.*;
import static javax.swing.JFrame.EXIT_ON_CLOSE;

public class Maze extends JPanel implements KeyListener, MouseListener {
	JFrame frame;
	String[][] maze;
	Explorer exp;
	boolean alreadyPainted = false;
	boolean alreadyRead = false;
	String[] lineArr = new String[33];
	int k = 0;
	String line;
	int lineCount = 0;
	boolean draw3D = false;
	int shrink = 50;
	ArrayList<Wall> walls;

	public Maze() {
		frame = new JFrame();
		frame.add(this);
		frame.setDefaultCloseOperation(EXIT_ON_CLOSE);
		frame.setSize(1600, 1000);
		frame.setVisible(true);
		frame.addKeyListener(this);
		frame.addMouseListener(this);
		try {
			BufferedReader reader = new BufferedReader(new FileReader("Classwork/dataStructures/src/mazeAttempts/Maze.txt"));
			if ((line = reader.readLine()) != null) {
				do {
					lineArr[k] = line;
					lineCount++;
					k++;
				} while ((line = reader.readLine()) != null);
			}
		} catch (Exception e) {
			System.out.println("Exception");
		}

	}

	public static void main(String[] args) {
		new Maze();
	}

	public void paintComponent(Graphics g) {
		super.paintComponent(g);
		Graphics2D g2 = (Graphics2D) (g);
		g.setColor(BLACK);
		g.fillRect(0, 0, 1600, 1000);
		k = 0;
		if (draw3D == false) {
			maze = new String[lineArr.length][lineArr[0].split("").length];
			for (int i = 0; i < lineArr.length; i++) {
				for (int j = 0; j < lineArr[0].split("").length; j++) {
					maze[i][j] = lineArr[i].split("")[j];
					if (maze[i][j].equals("#"))
						g.setColor(WHITE);
					else if (maze[i][j].equals("1") && alreadyPainted == false) {
						exp = new Explorer(RED, new Location(j, i), 15, 0);
						g.setColor(exp.getColor());
						alreadyPainted = true;
					} else if (maze[i][j].equals("E"))
						g.setColor(GREEN);
					else
						g.setColor(BLACK);
					g.fillRect(j * 15, i * 15, 15, 15);
				}
			}


			if (alreadyPainted == true) {
				g.setColor(exp.getColor());
				g2.fill(exp.getRectangle());
			}
		} else if (draw3D) {
			for (int i = 0; i < walls.size(); i++) {
				g2.setColor(WHITE);
				g2.fillPolygon(walls.get(i).getPolygon());
				g2.setColor(RED);
				g2.drawPolygon(walls.get(i).getPolygon());
			}
		}
	}

	public void createWalls() {
		walls = new ArrayList<Wall>();
		for (int n = 0; n < 5; n++) {
			walls.add(getCeil(n));
			walls.add(getFloor(n));
			//walls.add(getLeft(n));
			//walls.add(getRight(n));
		}
		int expR = exp.getLocation().getX();
		int expC = exp.getLocation().getY();
		int expDir = exp.getDir();
		/*
		 * code methods getRight and getLeft and also use a switch statement with directions to print walls on right, left, and front
		 * code the enhanced hallways: getRightPath, getRightFloor, getRightCeil
		 */
		switch (expDir) {
			case 0: //east
				for (int n = 0; n < 5; n++) {
					try {
						/*if (mazeAttempts.maze[expR - 1][expC - (5 - n)] == "#") {
							walls.add(getFront(5 - n));
						}*/
						if (maze[expC + 1][expR + n].equals("#")) {
							walls.add(getRight(n));
						}
						if (maze[expC - 1][expR + n].equals("#")) {
							walls.add(getLeft(n));
						}
					} catch (ArrayIndexOutOfBoundsException e) {
						e.printStackTrace();
					}
				}
				break;
			case 1: //south
				for (int n = 0; n < 5; n++) {
					try {
						//if(mazeAttempts.maze[expR-1][expC-(5-n)] == "#"){
						//	walls.add(getFront(5-n));
						//}
						if (maze[expC + n][expR - 1].equals("#")) {
							walls.add(getRight(n));
						}
						if (maze[expC + n][expR + 1].equals("#")) {
							walls.add(getLeft(n));
						}
					} catch (ArrayIndexOutOfBoundsException e) {
						e.printStackTrace();
					}
				}
				break;
			case 2: //west
				for (int n = 0; n < 5; n++) {
					try {
						//if(mazeAttempts.maze[expR-1][expC-(5-n)] == "#"){
						//	walls.add(getFront(5-n));
						//}
						if (maze[expC + 1][expR - n].equals("#")) {
							walls.add(getRight(n));
						}
						if (maze[expC - 1][expR - n].equals("#")) {
							walls.add(getLeft(n));
						}
					} catch (ArrayIndexOutOfBoundsException e) {
						e.printStackTrace();
					}
				}
				break;
			case 3: //north
				for (int n = 0; n < 5; n++) {
					try {
						//if(mazeAttempts.maze[expR-1][expC-(5-n)] == "#"){
						//	walls.add(getFront(5-n));
						//}
						if (maze[expC - n][expR + 1].equals("#")) {
							walls.add(getRight(n));
						}
						if (maze[expC - n][expR - 1].equals("#")) {
							walls.add(getLeft(n));
						}
					} catch (ArrayIndexOutOfBoundsException e) {
						e.printStackTrace();
					}
				}
				break;
			case -1: //north
				for (int n = 0; n < 5; n++) {
					try {
						//if(mazeAttempts.maze[expR-1][expC-(5-n)] == "#"){
						//	walls.add(getFront(5-n));
						//}
						if (maze[expC - n][expR + 1].equals("#")) {
							walls.add(getRight(n));
						}
						if (maze[expC - n][expR - 1].equals("#")) {
							walls.add(getLeft(n));
						}
					} catch (ArrayIndexOutOfBoundsException e) {
						e.printStackTrace();
					}
				}
				break;
			case -2: //west
				for (int n = 0; n < 5; n++) {
					try {
						//if(mazeAttempts.maze[expR-1][expC-(5-n)] == "#"){
						//	walls.add(getFront(5-n));
						//}
						if (maze[expC + 1][expR - n].equals("#")) {
							walls.add(getRight(n));
						}
						if (maze[expC - 1][expR - n].equals("#")) {
							walls.add(getLeft(n));
						}
					} catch (ArrayIndexOutOfBoundsException e) {
						e.printStackTrace();
					}
				}
				break;
			case -3: //south
				for (int n = 0; n < 5; n++) {
					try {
						//if(mazeAttempts.maze[expR-1][expC-(5-n)] == "#"){
						//	walls.add(getFront(5-n));
						//}
						if (maze[expC + n][expR - 1].equals("#")) {
							walls.add(getRight(n));
						}
						if (maze[expC + n][expR + 1].equals("#")) {
							walls.add(getLeft(n));
						}
					} catch (ArrayIndexOutOfBoundsException e) {
						e.printStackTrace();
					}
				}
				break;
		}
	}

	public Wall getFloor(int n) {
		int[] rLocs = {700 - n * shrink, 650 - n * shrink, 650 - n * shrink, 700 - n * shrink},
				cLocs = {100 + n * shrink, 150 + n * shrink, 650 - n * shrink, 700 - n * shrink};
		return new Wall(rLocs, cLocs, "Floor", 255 - n * shrink, 255 - n * shrink, 255 - n * shrink);
	}

	public Wall getCeil(int n) {
		int[] rLocs = {100 + n * shrink, 150 + n * shrink, 150 + n * shrink, 100 + n * shrink},
				cLocs = {100 + n * shrink, 150 + n * shrink, 650 - n * shrink, 700 - n * shrink};
		return new Wall(rLocs, cLocs, "Ceil", 255 - n * shrink, 255 - n * shrink, 255 - n * shrink);
	}

	public Wall getLeft(int n) {
		int[] rLocs = {100 + n * shrink, 150 + n * shrink, 650 - n * shrink, 700 - n * shrink},
				cLocs = {100 + n * shrink, 150 + n * shrink, 150 + n * shrink, 100 + n * shrink};
		return new Wall(rLocs, cLocs, "Left", 255 - n * shrink, 255 - n * shrink, 255 - n * shrink);
	}

	public Wall getRight(int n) {
		int[] rLocs = {100 + n * shrink, 150 + n * shrink, 650 - n * shrink, 700 - n * shrink},
				cLocs = {700 - n * shrink, 650 - n * shrink, 650 - n * shrink, 700 - n * shrink};
		return new Wall(rLocs, cLocs, "Right", 255 - n * shrink, 255 - n * shrink, 255 - n * shrink);
	}

	public Wall getFront(int n) {
		int[] rLocs = {100 + n * shrink, 100 + n * shrink, 700 - n * shrink, 700 - n * shrink},
				cLocs = {100 + n * shrink, 650 - n * shrink, 650 - n * shrink, 700 - n * shrink};
		return new Wall(rLocs, cLocs, "Ceil", 255 - n * shrink, 255 - n * shrink, 255 - n * shrink);
	}

	public void keyPressed(@NotNull KeyEvent e) {
		if (e.getKeyCode() == 32) draw3D = !draw3D;
		if (draw3D) createWalls();
		exp.move(e.getKeyCode(), maze, draw3D);
		repaint();
	}

	public void keyReleased(KeyEvent e) {

	}

	public void keyTyped(KeyEvent e) {

	}

	public void mouseClicked(MouseEvent e) {

	}

	public void mousePressed(MouseEvent e) {

	}

	public void mouseReleased(MouseEvent e) {

	}

	public void mouseEntered(MouseEvent e) {

	}

	public void mouseExited(MouseEvent e) {

	}
}

File: Classwork/Data_Structures/src/mazeAttempts/maze6/Maze.kt
/*
package mazeAttempts.maze6

*/
/*import javafx.application.Platform.*
import javafx.scene.media.Media
import javafx.scene.media.MediaPlayer*//*

import java.awt.Color
import java.awt.Graphics
import javax.swing.JPanel
import java.awt.event.KeyListener
import java.awt.event.MouseListener
import javax.swing.JFrame
import java.io.File
import java.io.BufferedReader
import java.io.FileReader
import java.io.IOException
import java.awt.Graphics2D
import java.awt.event.KeyEvent
import java.awt.event.MouseEvent
import java.lang.ArrayIndexOutOfBoundsException
import java.lang.Thread.*
import java.util.ArrayList
import kotlin.jvm.JvmStatic
import javax.swing.WindowConstants.*

class Maze1 : JPanel(), KeyListener, MouseListener {
	var frame: JFrame
	var list = ArrayList<String>()
	var walls2D = ArrayList<Location>()
	var walls = ArrayList<Wall>()
	var maze = Array(33) { CharArray(74) }
	var draw3D = false
	var explorerLocationIndex = 0
	var shrink = 50
	var counterM = 0
	var location = Location(2, 0)
	var explorer = Explorer(location, 0, 13, Color.WHITE)
	fun setBoard() {
		val name = File("Classwork/dataStructures/src/mazeAttempts/Maze.txt")
		try {
			val input = BufferedReader(FileReader(name))
			var txt: String
			while (input.readLine().also { txt = it } != null) {
				list.add(txt)
				for (i in 0 until txt.length) {
					maze[counterM][i] = txt[i]
				}
				counterM++
				for (i in 0 until txt.length) {
					if (txt[i] == '1') {
						explorerLocationIndex = Integer.valueOf(txt.indexOf('1'))
					}
				}
			}
		} catch (e: IOException) {
			e.printStackTrace()
		}
	}

	public override fun paintComponent(g: Graphics) {
		super.paintComponent(g)
		val g2 = g as Graphics2D
		if (!draw3D) {
			for (i in list.indices) {
				for (j in 0 until list[i].length) {
					val store = list[i].split("").toTypedArray()
					if (store[j] == "#") {
						g.setColor(Color.BLACK)
						g.fillRect(j * 13, i * 13, 13, 13)
						walls2D.add(explorer.location)
					} else if (store[j] == " ") {
						g.setColor(Color.GREEN)
						g.fillRect(j * 13, i * 13, 13, 13)
					} else if (store[j] == "1") {
						g.setColor(Color.GREEN)
						g.fillRect(j * 13, i * 13, 13, 13)
					} else if (store[j] == "E") {
						g.setColor(Color(255, 215, 0))
						g.fillRect(j * 13, i * 13, 13, 13)
					}
				}
			}
			g.setColor(explorer.color)
			g.fillRect(explorer.location.c * 13, explorer.location.r * 13, 13, 13)
		} else {
			for (w in walls) {
				if (w.type != "Front") {
					g2.paint = w.paint
					g2.fillPolygon(w.polygon)
					g2.color = w.color
					g2.drawPolygon(w.polygon)
				}
			}
			for (w in walls) {
				if (w.type == "Front") {
					g2.paint = w.color
					g2.fillPolygon(w.polygon)
					g2.color = w.color
					g2.drawPolygon(w.polygon)
				}
			}
		}
	}

	fun createWalls() {
		walls = ArrayList()
		walls.add(getFront(5))
		for (n in 0..4) {
			walls.add(getCeiling(n))
			walls.add(getFloor(n))
			walls.add(getLeftPath(n + 1))
			walls.add(getLeftCeil(n))
			walls.add(getLeftFloor(n))
			walls.add(getRightPath(n + 1))
			walls.add(getRightCeil(n))
			walls.add(getRightFloor(n))
		}
		val explorerR = explorer.location.r
		val explorerC = explorer.location.c
		val explorerDirection = explorer.direction
		when (explorerDirection) {
			0 -> {
				var n = 0
				while (n < 5) {
					try {
						if (maze[explorerR - (5 - n)][explorerC] == '#') {
							walls.add(getFront(5 - n))
						}
						if (maze[explorerR - n][explorerC + 1] == '#') {
							walls.add(getRight(n))
						}
						if (maze[explorerR - n][explorerC - 1] == '#') {
							walls.add(getLeft(n))
						}
					} catch (e: ArrayIndexOutOfBoundsException) {
					}
					n++
				}
			}
			1 -> {
				var n = 0
				while (n < 5) {
					try {
						if (maze[explorerR][explorerC + (5 - n)] == '#') {
							walls.add(getFront(5 - n))
						}
						if (maze[explorerR + 1][explorerC + n] == '#') {
							walls.add(getRight(n))
						}
						if (maze[explorerR - 1][explorerC + n] == '#') {
							walls.add(getLeft(n))
						}
					} catch (e: ArrayIndexOutOfBoundsException) {
					}
					n++
				}
			}
			2 -> {
				var n = 0
				while (n < 5) {
					try {
						if (maze[explorerR + (5 - n)][explorerC] == '#') {
							walls.add(getFront(5 - n))
						}
						if (maze[explorerR + n][explorerC - 1] == '#') {
							walls.add(getRight(n))
						}
						if (maze[explorerR + n][explorerC + 1] == '#') {
							walls.add(getLeft(n))
						}
					} catch (e: ArrayIndexOutOfBoundsException) {
					}
					n++
				}
			}
			3 -> {
				var n = 0
				while (n < 5) {
					try {
						if (maze[explorerR][explorerC - (5 - n)] == '#') {
							walls.add(getFront(5 - n))
						}
						if (maze[explorerR - 1][explorerC - n] == '#') {
							walls.add(getRight(n))
						}
						if (maze[explorerR + 1][explorerC + n] == '#') {
							walls.add(getLeft(n))
						}
					} catch (e: ArrayIndexOutOfBoundsException) {
					}
					n++
				}
			}
		}
	}

	fun getCeiling(n: Int): Wall {
		val rLocs = intArrayOf(100 + n * shrink, 150 + n * shrink, 150 + n * shrink, 100 + n * shrink)
		val cLocs = intArrayOf(100 + n * shrink, 150 + n * shrink, 650 - n * shrink, 700 - n * shrink)
		return Wall(rLocs, cLocs, "Ceil", 255 - n * shrink, 255 - n * shrink, 255 - n * shrink)
	}

	fun getFloor(n: Int): Wall {
		val rLocs = intArrayOf(700 - n * shrink, 650 - n * shrink, 650 - n * shrink, 700 - n * shrink)
		val cLocs = intArrayOf(100 + n * shrink, 150 + n * shrink, 650 - n * shrink, 700 - n * shrink)
		return Wall(rLocs, cLocs, "Floor", 255 - n * shrink, 255 - n * shrink, 255 - n * shrink)
	}

	fun getLeft(n: Int): Wall {
		val rLocs = intArrayOf(100 + n * shrink, 150 + n * shrink, 650 - n * shrink, 700 - n * shrink)
		val cLocs = intArrayOf(100 + n * shrink, 150 + n * shrink, 150 + n * shrink, 100 + n * shrink)
		return Wall(rLocs, cLocs, "Left", 255 - n * shrink, 255 - n * shrink, 255 - n * shrink)
	}

	fun getRight(n: Int): Wall {
		val rLocs = intArrayOf(100 + n * shrink, 150 + n * shrink, 650 - n * shrink, 700 - n * shrink)
		val cLocs = intArrayOf(700 - n * shrink, 650 - n * shrink, 650 - n * shrink, 700 - n * shrink)
		return Wall(rLocs, cLocs, "Right", 255 - n * shrink, 255 - n * shrink, 255 - n * shrink)
	}

	fun getFront(n: Int): Wall {
		val rLocs = intArrayOf(100 + n * shrink, 100 + n * shrink, 700 - n * shrink, 700 - n * shrink)
		val cLocs = intArrayOf(100 + n * shrink, 700 - n * shrink, 700 - n * shrink, 100 + n * shrink)
		return Wall(rLocs, cLocs, "Front", 255 - n * shrink, 255 - n * shrink, 255 - n * shrink)
	}

	fun getLeftPath(n: Int): Wall {
		val rLocs = intArrayOf(100 + n * shrink, 100 + n * shrink, 700 - n * shrink, 700 - n * shrink)
		val cLocs = intArrayOf(50 + n * shrink, 100 + n * shrink, 100 + n * shrink, 50 + n * shrink)
		return Wall(rLocs, cLocs, "LeftPath", 255 - n * shrink, 255 - n * shrink, 255 - n * shrink)
	}

	fun getLeftFloor(n: Int): Wall {
		val rLocs = intArrayOf(700 - n * shrink, 650 - n * shrink, 650 - n * shrink)
		val cLocs = intArrayOf(100 + n * shrink, 100 + n * shrink, 150 + n * shrink)
		return Wall(rLocs, cLocs, "LeftFloor", 255 - n * shrink, 255 - n * shrink, 255 - n * shrink)
	}

	fun getLeftCeil(n: Int): Wall {
		val rLocs = intArrayOf(100 + n * shrink, 150 + n * shrink, 150 + n * shrink)
		val cLocs = intArrayOf(100 + n * shrink, 150 + n * shrink, 100 + n * shrink)
		return Wall(rLocs, cLocs, "LeftCeil", 255 - n * shrink, 255 - n * shrink, 255 - n * shrink)
	}

	fun getRightPath(n: Int): Wall {
		val rLocs = intArrayOf(100 + n * shrink, 100 + n * shrink, 700 - n * shrink, 700 - n * shrink)
		val cLocs = intArrayOf(700 - n * shrink, 750 - n * shrink, 750 - n * shrink, 700 - n * shrink)
		return Wall(rLocs, cLocs, "RightPath", 255 - n * shrink, 255 - n * shrink, 255 - n * shrink)
	}

	fun getRightFloor(n: Int): Wall {
		val rLocs = intArrayOf(700 - n * shrink, 650 - n * shrink, 650 - n * shrink)
		val cLocs = intArrayOf(700 - n * shrink, 700 - n * shrink, 650 - n * shrink)
		return Wall(rLocs, cLocs, "RightFloor", 255 - n * shrink, 255 - n * shrink, 255 - n * shrink)
	}

	fun getRightCeil(n: Int): Wall {
		val rLocs = intArrayOf(100 + n * shrink, 150 + n * shrink, 150 + n * shrink)
		val cLocs = intArrayOf(700 - n * shrink, 700 - n * shrink, 650 - n * shrink)
		return Wall(rLocs, cLocs, "RightCeil", 255 - n * shrink, 255 - n * shrink, 255 - n * shrink)
	}

	override fun keyPressed(e: KeyEvent) {
		if (e.keyCode == 32) {
			draw3D = !draw3D
			println("Switch screens!")
		}
		if (e.keyCode == 37 || e.keyCode == 38 || e.keyCode == 39 || e.keyCode == 40) {
//			e.keyCode.move(list)
			if (explorer.direction == 0) println("Direction: north")
			if (explorer.direction == 1) println("Direction: east")
			if (explorer.direction == 2) println("Direction: south")
			if (explorer.direction == 3) println("Direction: west")
		}
		if (draw3D) {
			createWalls()
		}
		if (explorer.location.r == 30 && explorer.location.c == 72) {
			println("You win!")
		}
		repaint()
	}

	override fun keyReleased(e: KeyEvent) {}
	override fun keyTyped(e: KeyEvent) {}
	override fun mouseClicked(e: MouseEvent) {}
	override fun mousePressed(e: MouseEvent) {}
	override fun mouseReleased(e: MouseEvent) {}
	override fun mouseEntered(e: MouseEvent) {}
	override fun mouseExited(e: MouseEvent) {}

	companion object {
		@JvmStatic
		fun main(args: Array<String>) {
			val app = Maze()
			app.setBoard()
		}
	}

	init {
		frame = JFrame()
		frame.add(this)
		frame.defaultCloseOperation = EXIT_ON_CLOSE
		frame.setSize(1500, 800)
		frame.isVisible = true
		frame.addKeyListener(this)
		frame.addMouseListener(this)
		startup(currentThread())
		MediaPlayer(
			Media(
				File("Classwork/dataStructures/src/mazeAttempts/Background-Music.mp3").toURI().toString()
			)
		).play()
	}
}*/
